### PR TITLE
Facebook

### PR DESCRIPTION
**Pull request by ledamint**

### CODE REVIEW

File: packages/urban-bot-facebook/src/UrbanBotFacebook.ts [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+        app.use(json({ verify: this.verifyRequestSignature }));
+
+        app.get('/webhook', (req, res) => {
+            const mode = req.query['hub.mode'];
+            const token = req.query['hub.verify_token'];
+            const challenge = req.query['hub.challenge'];
+
+            if (mode && token) {
+                if (mode === 'subscribe' && token === this.options.verifyToken) {
+                    console.log('WEBHOOK_VERIFIED');
```
Comment (oTkPoBeHuE) [inline packages/urban-bot-facebook/src/UrbanBotFacebook.ts]: I think that  logging is not necessary in the right cases. Should you  use console.err only?

File: packages/urban-bot-facebook/src/UrbanBotFacebook.ts [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        const signature = req.headers['x-hub-signature'];
+
+        if (!signature) {
+            console.log("Couldn't validate the signature.");
+        } else {
+            const elements = signature.split('=');
+            const signatureHash = elements[1];
+            const expectedHash = crypto.createHmac('sha1', this.options.appSecret).update(buf).digest('hex');
+            if (signatureHash != expectedHash) {
+                throw new Error("Couldn't validate the request signature.");
```
Comment (oTkPoBeHuE) [inline packages/urban-bot-facebook/src/UrbanBotFacebook.ts:346]: Maybe: "Couldn't validate the request **hash**." ?

File: packages/urban-bot-facebook/src/UrbanBotFacebook.ts [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+                );
+            }
+        }
+    }
+
+    verifyRequestSignature = (req: any, _res: any, buf: any) => {
+        const signature = req.headers['x-hub-signature'];
+
+        if (!signature) {
+            console.log("Couldn't validate the signature.");
```
Comment (oTkPoBeHuE) [inline packages/urban-bot-facebook/src/UrbanBotFacebook.ts]: Should you use  console.err ?

File: packages/urban-bot-facebook/src/GraphAPI.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                userProfile[camelizedKey] = value;
+            }
+
+            return userProfile;
+        } catch (err) {
+            console.log('Fetch failed:', err);
+        }
+    }
+
+    callUserProfileAPI(senderPsid: any) {
```
Comment (oTkPoBeHuE) [inline packages/urban-bot-facebook/src/GraphAPI.ts]: Is _senderPsid_ a **string**?

File: packages/urban-bot-facebook/src/GraphAPI.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                access_token: this.options.pageAccessToken,
+                subscribed_fields: fields,
+            },
+            method: 'POST',
+        }).catch((error) => {
+            console.error('Unable to send message:', error);
+        });
+    }
+
+    async getUserProfile(senderPsid: any) {
```
Comment (oTkPoBeHuE) [inline packages/urban-bot-facebook/src/GraphAPI.ts]: Is _senderPsid_ a **string**?

File: packages/urban-bot-facebook/src/GraphAPI.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+                fields: fields,
+                include_values: 'true',
+            },
+            method: 'POST',
+        }).catch((error) => {
+            console.error('Unable to send message:', error);
+        });
+    }
+
+    callSubscribedApps(customFields: any) {
```
Comment (oTkPoBeHuE) [inline packages/urban-bot-facebook/src/GraphAPI.ts]: Is _customFields_ a **string**?

File: packages/urban-bot-facebook/src/GraphAPI.ts [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+ */
+
+import requestPromise from 'request-promise';
+import camelCase from 'camelcase';
+import { FacebookOptions } from './UrbanBotFacebook';
+
+export class GraphAPI {
+    constructor(public options: FacebookOptions) {}
+
+    callSendAPI(requestBody: any) {
```
Comment (oTkPoBeHuE) [inline packages/urban-bot-facebook/src/GraphAPI.ts:20]: Is _requestBody_ a **object** ?

**Review by oTkPoBeHuE — APPROVED** [review]

File: packages/urban-bot-facebook/src/GraphAPI.ts [hunk 8]
```diff
@@ -0,0 +1,209 @@
+/**
+ * Copyright 2019-present, Facebook, Inc. All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Messenger For Original Coast Clothing
+ * https://developers.facebook.com/docs/messenger-platform/getting-started/sample-apps/original-coast-clothing
+ */
```
Comment (oTkPoBeHuE) [inline packages/urban-bot-facebook/src/GraphAPI.ts:10]: Add link to file

File: packages/urban-bot-facebook/src/UrbanBotFacebook.ts (hunk 1 above)
Comment (ledamint) [reply packages/urban-bot-facebook/src/UrbanBotFacebook.ts]: This log is useful to understand that you set up your server right

File: packages/urban-bot-facebook/src/UrbanBotFacebook.ts (hunk 2 above)
Comment (ledamint) [reply packages/urban-bot-facebook/src/UrbanBotFacebook.ts:346]: I guess it is ok

File: packages/urban-bot-facebook/src/GraphAPI.ts (hunk 5 above)
Comment (ledamint) [reply packages/urban-bot-facebook/src/GraphAPI.ts]: Yes, fixed

File: packages/urban-bot-facebook/src/GraphAPI.ts (hunk 6 above)
Comment (ledamint) [reply packages/urban-bot-facebook/src/GraphAPI.ts]: yes, fixed

File: packages/urban-bot-facebook/src/GraphAPI.ts (hunk 7 above)
Comment (ledamint) [reply packages/urban-bot-facebook/src/GraphAPI.ts:20]: Yes, I will specify it later

File: packages/urban-bot-facebook/src/GraphAPI.ts (hunk 4 above)
Comment (ledamint) [reply packages/urban-bot-facebook/src/GraphAPI.ts]: Fixed

File: packages/urban-bot-facebook/src/GraphAPI.ts (hunk 8 above)
Comment (ledamint) [reply packages/urban-bot-facebook/src/GraphAPI.ts:10]: Added

**Pull request merged**